nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java

```diff
@@ -0,0 +1,283 @@
+import org.sonatype.nexus.repository.storage.Query;
+import org.sonatype.nexus.repository.storage.StorageTx;
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.repository.view.Payload;
+import org.sonatype.nexus.repository.view.payloads.BlobPayload;
+
+import static java.util.Collections.singletonList;
+import static org.sonatype.nexus.common.hash.HashAlgorithm.SHA1;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.ARTIFACT_JAR;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.ARTIFACT_XML;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.ARTIFACT_XML_XZ;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.COMPONENT_BINARY;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.COMPONENT_FEATURES;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.COMPONENT_PLUGINS;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.COMPOSITE_ARTIFACTS_JAR;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.COMPOSITE_ARTIFACTS_XML;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.COMPOSITE_CONTENT_JAR;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.COMPOSITE_CONTENT_XML;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.CONTENT_JAR;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.CONTENT_XML;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.CONTENT_XML_XZ;
+import static org.sonatype.nexus.repository.p2.internal.AssetKind.P2_INDEX;
+import static org.sonatype.nexus.repository.p2.internal.proxy.P2ProxyRecipe.ARTIFACTS_NAME;
+import static org.sonatype.nexus.repository.p2.internal.proxy.P2ProxyRecipe.COMPOSITE_ARTIFACTS;
+import static org.sonatype.nexus.repository.p2.internal.proxy.P2ProxyRecipe.COMPOSITE_CONTENT;
+import static org.sonatype.nexus.repository.p2.internal.proxy.P2ProxyRecipe.CONTENT_NAME;
+import static org.sonatype.nexus.repository.p2.internal.proxy.P2ProxyRecipe.JAR_EXTENSION;
+import static org.sonatype.nexus.repository.p2.internal.proxy.P2ProxyRecipe.XML_EXTENSION;
+import static org.sonatype.nexus.repository.p2.internal.proxy.P2ProxyRecipe.XML_XZ_EXTENSION;
+import static org.sonatype.nexus.repository.p2.internal.util.P2PathUtils.PLUGIN_NAME;
+import static org.sonatype.nexus.repository.storage.AssetEntityAdapter.P_ASSET_KIND;
+import static org.sonatype.nexus.repository.storage.ComponentEntityAdapter.P_VERSION;
+import static org.sonatype.nexus.repository.storage.MetadataNodeEntityAdapter.P_NAME;
+
+/**
+ * {@link P2Facet} implementation
+ *
+ * @since 0.next
+ */
+@Named
+public class P2FacetImpl
+    extends FacetSupport
+    implements P2Facet
+{
+  public static final List<HashAlgorithm> HASH_ALGORITHMS = ImmutableList.of(SHA1);
+
+  @Override
+  public Component findOrCreateComponent(final StorageTx tx, final String path, final Map<String, String> attributes) {
```

Comment:
we already have appropriate DTO P2Attributes. Why used a hashmap ?

Reply:
We already have doCreateOrSaveComponent in P2ProxyFacetImpl. 
findOrCreateComponent and findOrCreateAsset looks like a code duplication

Reply:
P2ProxyFacet is not visible from the module, the same for P2Attributes. Also, doCreateOrSaveComponent have as parameter Payload which we don't have on this case